### PR TITLE
Fix 3

### LINKED ISSUE #3: Implement ensemble analyses
### Description

Some ideas:

- Ensemble statistics (ensemble mean, spread, change vs. historical, etc.). A lot already exists in `xclim.ensemble`.
- Projections in T° change vs. preindustrial levels

Here too, most of the work should be done within `xclim`, with `xscen` acting as a wrapper.

### PR DESCRIPTION
**Pull request by juliettelavoie**
### Pull Request Checklist:
- [x] `pre-commit` hooks are installed/active in my local clone (`$ pre-commit install`)
- [x] This PR addresses an already opened issue (for bug fixes / features)
    - This PR fixes #3
- [x] (If applicable) Documentation has been added / updated (for bug fixes / features)
- [x] If a merge request has been made in parallel to this PR in xscen-notebooks, it is merged and the submodules have been updated.
- [x] HISTORY.rst has been updated (with summary of main changes)
  - [x] Link to issue (:issue:`number`) and pull request (:pull:`number`) has been added
- [ ] `bumpversion patch` has been called on this branch

### What kind of change does this PR introduce?

- new function `ensemble_stats` 
  -  It creates an ensemble and call a stats function from xclim.
  - option 'common_attrs_only` deletes all attrs that are not common to all datasets in the ensemble
- ensemble section added to the notebook. MR: (https://gitlab.ouranos.ca/scenario/xscen-notebooks/merge_requests/2) 
  -  It seems like I can merge the MR without review. I will consider that you accept that MR if you accept this PR. 
- I have also made small change to the handling of attrs in `train` (to be able to use .to_dataset_dict() directly) and `regrid` (to not fail if the input ds does not come from a catalog)

### Does this PR introduce a breaking change?

no

### CODE REVIEW

File: xscen/ensemble.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        Dataset with ensemble statistics
+    """
+    logger.info(f"Creating ensemble with {len(datasets)} and calculating {statistics}.")
+
+    # if input files are .zarr, change the engine automatically
+    if isinstance(datasets[0], (str, Path)):
+        path = Path(datasets[0])
+        if path.suffix == ".zarr" and "engine" not in create_kwargs:
+            create_kwargs["engine"] = "zarr"
+
```
Comment (RondeauG) [inline xscen/ensemble.py:55]: I would add:
```
if (len(datasets) > 1) and ():
   create_kwargs.setdefault({"mf_flag": True})
```
Under the `if isinstance (...)`

File: xscen/ensemble.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    ens = ensembles.create_ensemble(datasets, **create_kwargs)
+    ens_stats = getattr(ensembles, statistics)(ens, **stats_kwargs)
+
+    # delete attrs that are not common to all dataset
+    if common_attrs_only:
+        for i in range(len(datasets)):
+            if isinstance(datasets[0], (str, Path)):
+                ds = xr.open_dataset(datasets[i], **create_kwargs)
```
Comment (RondeauG) [inline xscen/ensemble.py:63]: That will not work if someone calls kwargs specific to `create_ensemble` such as `resample_freq`.

File: xscen/ensemble.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        if path.suffix == ".zarr" and "engine" not in create_kwargs:
+            create_kwargs["engine"] = "zarr"
+
+    ens = ensembles.create_ensemble(datasets, **create_kwargs)
+    ens_stats = getattr(ensembles, statistics)(ens, **stats_kwargs)
+
+    # delete attrs that are not common to all dataset
+    if common_attrs_only:
+        for i in range(len(datasets)):
+            if isinstance(datasets[0], (str, Path)):
```
Comment (RondeauG) [inline xscen/ensemble.py:62]:
```suggestion
            if isinstance(datasets[i], (str, Path)):
```

File: xscen/ensemble.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+    Returns
+    -------
+    ens_stats: xr.Dataset
+        Dataset with ensemble statistics
+    """
+    logger.info(f"Creating ensemble with {len(datasets)} and calculating {statistics}.")
+
+    # if input files are .zarr, change the engine automatically
+    if isinstance(datasets[0], (str, Path)):
```
Comment (RondeauG) [inline xscen/ensemble.py:51]: Should we have a check that all entries are the same type?

File: xscen/ensemble.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+                    del ens_stats.attrs[a_key]
+        # create dataframe of catalogue attrs to generate new id
+        df = pd.DataFrame.from_dict(
+            {
+                key[4:]: [value]
+                for key, value in ens_stats.attrs.items()
+                if key[:4] == "cat/"
+            }
+        )
+        ens_stats.attrs["cat/id"] = generate_id(df)[0]
```
Comment (RondeauG) [inline xscen/ensemble.py:82]: Have you tested this? 
What's the ID created when you have multiple models & does it make sense?

File: xscen/regridding.py [hunk 6]
```diff
@@ -146,7 +146,9 @@ def regrid(
 
     # Attrs
     out.attrs["cat/processing_level"] = to_level
-    out.attrs["cat/domain"] = ds_grid.attrs["cat/domain"]
+    out.attrs["cat/domain"] = (
```
Comment (RondeauG) [inline xscen/regridding.py:149]: We'll have to do that change elsewhere too. I know that `indicators.compute_indicator` crashes for that same reason if `cat/variable` is not there...

**Review by RondeauG — COMMENTED** [review]
Looks good!

**Comment by RondeauG** [comment]
Do we want `xscen.ensembles` (plural), to fit with `xclim.ensembles`?

File: xscen/biasadjust.py [hunk 7]
```diff
@@ -160,6 +160,11 @@ def train(
         "jitter_over": jitter_over,
     }
 
+    # attrs that are needed to open with .to_dataset_dict()
```
[the remark this answers is not in the document]
Comment (juliettelavoie) [reply xscen/biasadjust.py:163]: Pour ESPO-G je sauve le `ds_train` en zarr et dans le catalogue (avec ces infos dans `info_dict` de `update_from_ds`) et je le réouvre pour adjust. donc, ce n'est pas absolument nécessaire, mais je ne pense pas que ça fait de mal.

File: xscen/ensemble.py (hunk 2 above)
Comment (juliettelavoie) [reply xscen/ensemble.py:63]: ahh you're right! Je vais changer ça.

File: xscen/ensemble.py (hunk 4 above)
Comment (juliettelavoie) [reply xscen/ensemble.py:51]: est-ce que ça va vraiment arrivé de passé un liste mélangée de paths et datasets?

File: xscen/ensemble.py (hunk 2 above)
Comment (RondeauG) [reply xscen/ensemble.py:63]: Would this work? `**create_kwargs["xr_kwargs"] if "xr_kwargs" in create_kwargs else {}`

File: xscen/ensemble.py (hunk 5 above)
Comment (juliettelavoie) [reply xscen/ensemble.py:82]: Si tu as plusieurs modèles, 'cat/source' serait effacer alors il ne serait pas inclus dans le id. J'ai testé avec mes fichiers info-crue et ça donne : 'info-crue-CMIP6_CMIP6_ScenarioMIP_ssp245_qc' (info-crue-CMIP6 == bias_adjusted_project)

Comment (RondeauG) [reply xscen/ensemble.py:82]: Ouais ok, looks good enough.

File: xscen/ensemble.py (hunk 4 above)
Comment (RondeauG) [reply xscen/ensemble.py:51]: Ouin ok, je vois que `xclim.ensembles.create_ensemble` ne fait pas de checks non plus. J'imagine que ça va juste planter si quelqu'un essaie d'envoyer un mélange des deux.

File: xscen/ensemble.py (hunk 1 above)
Comment (juliettelavoie) [reply xscen/ensemble.py:55]: je ne comprends pas le `and ()` ?

Comment (RondeauG) [reply xscen/ensemble.py:55]: Oups, tu peux oublier. C'était supposé être `isinstance(datasets[0], (str, Path))`, mais je me suis rendu compte que c'était déjà là. J'ai juste oublié d'enlever.

File: xscen/regridding.py (hunk 6 above)
Comment (juliettelavoie) [reply xscen/regridding.py:149]: J'ai fait le tour des "cat/" et modifié 2 endroits  pour qu'ils ne fassent plus planter la fonction en leur absence.

File: xscen/ensemble.py (hunk 2 above)
Comment (juliettelavoie) [reply xscen/ensemble.py:63]: non parce que xr_kwargs n'est pas un kwargs de `create`, ça fait juste représenter les kwargs additionnels (genre `engine`)

**Pull request merged**